### PR TITLE
drivers: udc_nrf: handle overwritten Set Address commands

### PR DESCRIPTION
**Pull request by tmon-nordic**
USBD peripheral automatically handles Set Address command which can lead to state mismatch between USB stack and the host. Keep track of device address and issue fake Set Address commands on mismatch.

This fixes default vs addressed state mismatch that can occur due to sufficently high SETUP handling latency. The state mismatch was most commonly seen as SET CONFIGURATION failure when the enumeration happened during periods with increased latency.

### CODE REVIEW

File: drivers/usb/udc/udc_nrf.c [hunk 1]
```diff
@@ -328,7 +332,77 @@ static int udc_event_xfer_setup(const struct device *dev)
...
+	/* USBD peripheral automatically handles Set Address in slightly
+	 * different manner than the USB stack.
+	 *
+	 * USBD peripheral doesn't care about wLength, but the peripheral
+	 * switches to new address only after status stage. The device won't
+	 * automatically accept Data Stage packets.
+	 *
+	 * However, in the case the host:
+	 *   * sends SETUP Set Address with non-zero wLength
+	 *   * does not set corresponding OUT DATA packets (to match wLength)
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_nrf.c:347]: Should it be "does not **send** corresponding OUT DATA packets (to match wLength)"?

File: drivers/usb/udc/udc_nrf.c [hunk 2]
```diff
@@ -328,7 +332,77 @@ static int udc_event_xfer_setup(const struct device *dev)
...
+	 *     or sends the packets but disregards NAK
+	 *     or sends the packets that device ACKs
+	 *   * sends IN token (either incorrectly proceeds to status stage, or
+	 *     manages to send IN before SW sets STALL)
+	 * then the USBD peripheral will accept the address and USB stack won't.
+	 * This will lead to state mismatch between the stack and peripheral.
+	 *
+	 * In cases where the SW stack would like to STALL the request there is
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_nrf.c:355]: "In cases where the USB stack would" or "In cases where the software would"?

File: drivers/usb/udc/udc_nrf.c (hunk 1 above)
Comment (tmon-nordic) [reply drivers/usb/udc/udc_nrf.c]: Fixed.

File: drivers/usb/udc/udc_nrf.c (hunk 2 above)
Comment (tmon-nordic) [reply drivers/usb/udc/udc_nrf.c]: Went with the first option.

**Review by jfischer-no — APPROVED** [review]

**Comment by dkalowsk** [comment]
@carlescufi and @tmon-nordic is this really needing to be in 4.0?  it doesn't have enough approvals and RC3 is already done.  I'd like move to 4.1.0 and pull it into a 4.0.1 backport if it is that needed.

**Review by MarekPieta — APPROVED** [review]
Thank you for the fix. I confirmed locally that issues reported by me no longer replicate

**Review by carlescufi — APPROVED** [review]

**Pull request merged**